sympy/ntheory/generate.py

```diff
@@ -421,23 +430,45 @@ def prime(nth):
     n = as_int(nth)
     if n < 1:
         raise ValueError("nth must be a positive integer; prime(1) == 2")
+
+    # Check if n is within the sieve range
     if n <= len(sieve._list):
         return sieve[n]
 
     from sympy.functions.elementary.exponential import log
     from sympy.functions.special.error_functions import li
-    a = 2 # Lower bound for binary search
-    # leave n inside int since int(i*r) != i*int(r) is not a valid property
-    # e.g. int(2*.5) != 2*int(.5)
-    b = int(n*(log(n) + log(log(n)))) # Upper bound for the search.
+    # Use approximations for smaller values of n to avoid li computation
+    if n < 1000:  # Threshold for approximation can be adjusted
+        # Extend sieve up to 8*n as this is empirically sufficient
+        sieve.extend(8 * n)
+        return sieve[n]
 
+    # Dynamically extend the sieve up to an estimated upper bound
+    upper_bound = int(n * (log(n).evalf() + log(log(n)).evalf()))
+
+    # Lower bound for binary search
+    a = 2
+
+    # Binary search for the least m such that li(m) > n
+    b = upper_bound
     while a < b:
         mid = (a + b) >> 1
-        if li(mid) > n:
+        li_mid = li(mid).evalf()
+        if li_mid > n:
             b = mid
         else:
             a = mid + 1
-    return nextprime(a - 1, n - _primepi(a - 1))
+
+    # Compute the number of primes up to m-1
+    n_primes = _primepi(a - 1)
+
+    # Find remaining primes starting from a
+    while n_primes < n:
+        if isprime(a):  # Use isprime() function instead of sieve.isprime
+            n_primes += 1
+        a += 1
+
+    return a - 1
```

Comment:
The code has regressed. It was fixed in #27035.

Reply:
Understood, 

Implemented the following:

```python

# Binary search for the least m such that li(m) > n
    while a < b:
        mid = (a + b) >> 1
        if li(mid).evalf() > n:
            b = mid
        else:
            a = mid + 1

    # Use nextprime to efficiently find the nth prime
    return nextprime(a - 1, n - _primepi(a - 1))```
    
    Works well.